scala/sexy-primes-test.scala

```diff
@@ -0,0 +1,245 @@
+    is_prime = (n: Int) => {
+      if ((n & 1) == 0) {
+        false
+      } else {
+        var i = 3
+        var result = 1
+        breakable {
+          while (true) {
+            if ((n % i) == 0) {
+              result = 0
+              break()
+            }
+            if (i * i > n) break()
+            i += 2
+          }
+        }
+        result != 0
+      }
+    }
+  }
+
+  else if (t == "orgm") {
+    is_prime = (n: Int) => {
+      (n & 1) != 0 && (3 until(sqrt(n).intValue + 1, 2)).forall(i => (n % i) != 0)
+    }
+  }
+
+  else if (t == "siev") {
+    primes_sieve = (n: Int) => {
+      // temporary "half" mask sieve for primes < n (using bool)
+      var sieve = Array.fill(n / 2)(true)
+      for (i <- 3 until(sqrt(n).intValue + 1, 2)) {
+        if (sieve(i / 2)) {
+          (i * i / 2 until(sieve.size, i)).foreach(x => {
+            sieve(x) = false
+          })
+        }
+      }
+      sieve
+    }
+    sexy_primes_below = Some((n: Int) => {
+      val sieve = primes_sieve(n + 1)
+      for (
+        j <- 9 until n + 1;
+        i = j - 6;
+        if ((i & 1) != 0 && sieve(i / 2) && sieve(j / 2))
+      ) yield Array(i, j)
+    })
+  }
+
+  else if (List("osie1", "osie2").contains(t)) {
+    var primesSieveInt: (Int) => Array[Int] = null
+    if (t == "osie1") {
+      primesSieveInt = (n: Int) => {
+        // temporary odd direct sieve for primes < n
+        val sieve = (3 until(n, 2)).toArray
+        val l = sieve.length
+        breakable {
+          for (i <- sieve) {
+            if (i != 0) {
+              val f = (i * i - 3) / 2
+              if (f >= l)
+                break()
+              (f until(sieve.size, i)).foreach({
+                sieve(_) = 0
+              })
+            }
+          }
+        }
+        sieve
+      }
+    }
+    if (t == "osie2") {
+      primesSieveInt = (n: Int) => {
+        // temporary odd direct sieve for primes < n
+        val l = ((n - 3) / 2)
+        val sieve = Array.fill(l)(-1)
+        breakable {
+          for (k <- 3 until(n, 2)) {
+            val o = (k - 3) / 2
+            var i = sieve(o)
+            if (i == -1) {
+              i = k
+              sieve((k - 3) / 2) = k
+            }
+            if (i != 0) {
+              val f = (i * i - 3) / 2
+              if (f >= l)
+                break()
+              (f until(sieve.size, i)).foreach({
+                sieve(_) = 0
+              })
+            }
+          }
+        }
+        sieve
+      }
+    }
+    sexy_primes_below = Some((n: Int) => {
+      val sieve = primesSieveInt(n + 1)
+      for (
+        j <- 9 until n + 1;
+        i = j - 6;
+        if ((i & 1) != 0 && sieve((i - 2) / 2) != 0 && sieve((j - 2) / 2) != 0)
+      ) yield Array(i, j)
+    })
+  }
+
+  else
+    proceed = false
+
+  if (proceed) {
+    // simple "sexy_primes_below" using is_prime :
+    if (!sexy_primes_below.isDefined) {
+      sexy_primes_below = Some((n: Int) => {
+        for (
```

Comment:
I believe, we can make that more faster, if we use a step 2 here for j iteration, for python also - would be `for j in range(9, n+1, 2):`. But then it would be not the original code.